### PR TITLE
Store unconnected payloads in Mempool

### PR DESCRIPTION
**Pull request by Mr-Leshiy**

### CODE REVIEW

File: include/veriblock/blockchain/temp_block_tree.hpp [hunk 1]
```diff
@@ -53,13 +61,39 @@ struct TempBlockTree {
...
+
+    if (index != nullptr) {
+      removeTempSingleBlock(*index);
+    }
+  }
+
+  void removeTempSingleBlock(index_t& index) {
+    // TODO: decrease counter for the previously known blocks
+    auto shortHash = tree_->makePrevHash(index.getHash());
+    auto it = temp_blocks_.at(shortHash);
```
Comment (blackyblack) [inline include/veriblock/blockchain/temp_block_tree.hpp:75]: Assert that temp_blocks_ contains shortHash

File: src/blockchain/mempool_block_tree.cpp [hunk 2]
```diff
@@ -1,49 +1,144 @@
...
+  }
+
+  auto endorsed_hash =
+      tree_->getParams().getHash(atv.transaction.publicationData.header);
+  auto* endorsed_index = tree_->getBlockIndex(endorsed_hash);
+  if (endorsed_index != nullptr) {
+    auto* tip = tree_->getBestChain().tip();
+    assert(tip != nullptr && "block tree is not bootstrapped");
+
+    if (tip && (tip->getHeight() - endorsed_index->getHeight() + 1 > window)) {
```
Comment (blackyblack) [inline src/blockchain/mempool_block_tree.cpp:34]: No need to check tip "if (tip)" here.
Add assert that endorsed_index height is not higher than tip height. Or avoid subtracting like: tip->getHeight() + 1 > window + endorsed_index->getHeight()

File: src/blockchain/mempool_block_tree.cpp [hunk 3]
```diff
@@ -1,49 +1,144 @@
...
+  if (duplicate) {
+    return state.Invalid(
+        "vtb-duplicate",
+        fmt::sprintf("VTB=%s already added to active chain in block %s",
+                     vtb.getId().toHex(),
+                     duplicate->toShortPrettyString()));
+  }
+
+  if (vtb.containingBlock.height - vtb.transaction.publishedBlock.height >
+      window) {
```
Comment (blackyblack) [inline src/blockchain/mempool_block_tree.cpp:66]: Same concern about heights subtraction.

File: include/veriblock/mempool.hpp [hunk 4]
```diff
@@ -87,12 +87,40 @@ struct MemPool {
...
+   *
+   * @note When payloads are received by rpc, you can do contextual validation
+   * immediately. If payloads are added during reorg/p2p sync, you can disable
+   * contextual validation, as otherwise it may ban valid payloads from remote
+   * peers.
+   *
+   * @tparam shared_ptr<T> one of VTB, ATV, VbkBlock
+   * @param[in] pl payload
+   * @param[out] state validation state
+   * @param[in] shouldDoContextualCheck if true, mempool performs contextual
```
[the remark this answers is not in the document]
Comment (Mr-Leshiy) [reply include/veriblock/mempool.hpp:112]: fixed

File: include/veriblock/blockchain/temp_block_tree.hpp (hunk 1 above)
Comment (Mr-Leshiy) [reply include/veriblock/blockchain/temp_block_tree.hpp:75]: added

File: include/veriblock/mempool.hpp [hunk 5]
```diff
@@ -87,12 +87,40 @@ struct MemPool {
...
    */
   template <typename T>
-  bool submit(const T& pl,
+  bool submit(const T& pl, ValidationState& state) {
+    return submit<T>(std::make_shared<T>(pl), state);
+  }
+
+  //! submit new payload into mempool
+  //! has overloads for VTB, ATV, VbkBlock
+  //! @ingroup api
```
[the remark this answers is not in the document]
Comment (Mr-Leshiy) [reply include/veriblock/mempool.hpp:96]: fixed

File: src/blockchain/mempool_block_tree.cpp (hunk 2 above)
Comment (Mr-Leshiy) [reply src/blockchain/mempool_block_tree.cpp:34]: fixed

File: src/blockchain/mempool_block_tree.cpp (hunk 3 above)
Comment (Mr-Leshiy) [reply src/blockchain/mempool_block_tree.cpp]: fixed

File: include/veriblock/mempool.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Mr-Leshiy) [reply include/veriblock/mempool.hpp]: fixed

**Review by Warchant — APPROVED** [review]

**Pull request merged**